lib/spack/spack/cmd/checksum.py

```diff
@@ -56,33 +56,29 @@ def checksum(parser, args):
     # Get the package we're going to generate checksums for
     pkg = spack.repo.get(args.package)
 
-    url_dict = {}
+    # Otherwise, see what versions we can find online
+    url_dict = pkg.fetch_remote_versions()
```

Comment:
Before this was done only conditionally (no args given) because it can be extremely expensive (e.g. `qt` which does a lot of spidering). What's the runtime of `spack checksum qt@5.nonexistent` on this branch?

Reply:
It is expensive, but it's also the only way to get correct behavior for packages where the URLs generated by their packages may or may not be correct.  The direct answer to your question is that `qt@5.nonexistent` takes 3 minutes and 43 seconds.

As far as I can think right now, the only way to deal with this better is to make it illegal for a package to generate broken links, but then R packages can't work.  We could work around that by changing the contract for generating URLs from a version to produce a list of URLs, at least one of which must be valid if the version is valid, but that would be a breaking change.

I suppose one intermediate step might be that we can do a check on the provided url and see if we get a 404, and only then fetch remote versions if we hit a failure from the package generated url?

Reply:
Yuck. Does it take that long to start downloading a valid version, such as 5.15.2? I know the qt "list URLs" is a bit of an outlier but it's going to be ugly if a new version comes out but takes minutes to verify the URL, even if we know it exists.

Reply:
Yes, it would be ideal if you could make `pkg.fetch_remote_versions` a lazy evaluation... it's something of a nuclear option, and because it uses the URL spidering can be unstable -- best if the checksum command only calls it if really necessary.

Reply:
Ok, let me prototype something.  It will have to be a bigger change, and there will still be unconditional requests, but cheaper ones that don't involve spidering by default.

Reply:
Thanks @trws !

Reply:
> As far as I can think right now, the only way to deal with this better is to make it illegal for a package to generate broken links, but then R packages can't work.

The R package locations are actually very predictable. The spidering should not be too bad as each package has a directory under "Archive", the `list_url` points to it, and there should not be that much in those respective directories. The location of an older source package is actually known but there is no mechanism in spack to fetch it other than `list_url`, at least none that I am aware of.

This may be out of scope for this PR but would it be feasible to have a second `url` attribute in spack? An R package could then look something like this:
```
url = "https://cloud.r-project.org/src/contrib/abind_1.4-5.tar.gz"
archive_url = "https://cloud.r-project.org/src/contrib/Archive/abind/abind_1.4-5.tar.gz"
```

The above would be handled by the `cran` attribute as `url`, `list_url`, and `homepage` are currently. Basically, if the source is not found at `url` try `archive_url`, using the same logic for replacing version numbers.

Reply:
Ok, so actually it should use the `urls` attribute rather than the `url` attribute, that will let both be specified and both will be tried.  The way it is now, with one as `list_url` will also work after this change.

Reply:
Ok now that's interesting.  The change I had but walked back would fix that, I expected current behavior would as well.  I'm tempted to say keep it list_url for now, and we can open something to discuss whether the fetch strategy should use my new logic for using fetch checks by default.  Honestly I think it should, but there's a unit test that walks every version in every package, and generates fetch strategies for them.  That is a recipe for disaster so I walked it back to use just the new url listing stuff, but it should check all the mirrors rather than just dying... that seems like a separate issue though since it dies currently too I guess.